### ISSUE TITLE
Errors building for node 11.8

### ISSUE THREAD

**Issue by trxcllnt** [holepunchto/sodium-native#87]
Getting the following errors building on node 11 (on aarch64):

```
In file included from ../binding.cc:12:0:
../src/crypto_pwhash_async.cc: In constructor ‘CryptoPwhashAsync::CryptoPwhashAsync(Nan::Callback*, unsigned char*, long long unsigned int, const char*, long long unsigned int, const unsigned char*, long long unsigned int, size_t, int)’:
../src/crypto_pwhash_async.cc:9:195: error: no matching function for call to ‘Nan::AsyncWorker::AsyncWorker(Nan::Callback*&, const char [34])’
 cWorker(callback, "sodium-native:crypto_pwhash_async"), out(out), outlen(outlen), passwd(passwd), passwdlen(passwdlen), salt(salt), opslimit(opslimit), memlimit(memlimit), alg(alg) {}
                                                                                                                                                                                    ^
In file included from ../binding.cc:3:0:
../../nan/nan.h:1506:12: note: candidate: Nan::AsyncWorker::AsyncWorker(Nan::Callback*)
   explicit AsyncWorker(Callback *callback_)
            ^~~~~~~~~~~
../../nan/nan.h:1506:12: note:   candidate expects 1 argument, 2 provided
In file included from ../binding.cc:12:0:
../src/crypto_pwhash_async.cc: In member function ‘virtual void CryptoPwhashAsync::HandleOKCallback()’:
../src/crypto_pwhash_async.cc:23:29: error: ‘async_resource’ was not declared in this scope
     callback->Call(1, argv, async_resource);
                             ^~~~~~~~~~~~~~
../src/crypto_pwhash_async.cc: In member function ‘virtual void CryptoPwhashAsync::HandleErrorCallback()’:
../src/crypto_pwhash_async.cc:33:29: error: ‘async_resource’ was not declared in this scope
     callback->Call(1, argv, async_resource);
                             ^~~~~~~~~~~~~~
In file included from ../binding.cc:13:0:
../src/crypto_pwhash_str_async.cc: In constructor ‘CryptoPwhashStrAsync::CryptoPwhashStrAsync(Nan::Callback*, char*, const char*, long long unsigned int, long long unsigned int, size_t)’:
../src/crypto_pwhash_str_async.cc:9:161: error: no matching function for call to ‘Nan::AsyncWorker::AsyncWorker(Nan::Callback*&, const char [38])’
     : Nan::AsyncWorker(callback, "sodium-native:crypto_pwhash_str_async"), out(out), passwd(passwd), passwdlen(passwdlen), opslimit(opslimit), memlimit(memlimit) {}
                                                                                                                                                                 ^
In file included from ../binding.cc:3:0:
../../nan/nan.h:1506:12: note: candidate: Nan::AsyncWorker::AsyncWorker(Nan::Callback*)
   explicit AsyncWorker(Callback *callback_)
            ^~~~~~~~~~~
../../nan/nan.h:1506:12: note:   candidate expects 1 argument, 2 provided
In file included from ../binding.cc:13:0:
../src/crypto_pwhash_str_async.cc: In member function ‘virtual void CryptoPwhashStrAsync::HandleOKCallback()’:
../src/crypto_pwhash_str_async.cc:23:29: error: ‘async_resource’ was not declared in this scope
     callback->Call(1, argv, async_resource);
                             ^~~~~~~~~~~~~~
../src/crypto_pwhash_str_async.cc: In member function ‘virtual void CryptoPwhashStrAsync::HandleErrorCallback()’:
../src/crypto_pwhash_str_async.cc:33:29: error: ‘async_resource’ was not declared in this scope
     callback->Call(1, argv, async_resource);
                             ^~~~~~~~~~~~~~
In file included from ../binding.cc:14:0:
../src/crypto_pwhash_str_verify_async.cc: In constructor ‘CryptoPwhashStrVerifyAsync::CryptoPwhashStrVerifyAsync(Nan::Callback*, const char*, const char*, long long unsigned int)’:
../src/crypto_pwhash_str_verify_async.cc:9:128: error: no matching function for call to ‘Nan::AsyncWorker::AsyncWorker(Nan::Callback*&, const char [45])’
     : Nan::AsyncWorker(callback, "sodium-native:crypto_pwhash_str_verify_async"), str(str), passwd(passwd), passwdlen(passwdlen) {}
                                                                                                                                ^
In file included from ../binding.cc:3:0:
../../nan/nan.h:1506:12: note: candidate: Nan::AsyncWorker::AsyncWorker(Nan::Callback*)
   explicit AsyncWorker(Callback *callback_)
            ^~~~~~~~~~~
../../nan/nan.h:1506:12: note:   candidate expects 1 argument, 2 provided
In file included from ../binding.cc:14:0:
../src/crypto_pwhash_str_verify_async.cc: In member function ‘virtual void CryptoPwhashStrVerifyAsync::HandleOKCallback()’:
../src/crypto_pwhash_str_verify_async.cc:27:29: error: ‘async_resource’ was not declared in this scope
     callback->Call(2, argv, async_resource);
                             ^~~~~~~~~~~~~~
../src/crypto_pwhash_str_verify_async.cc: In member function ‘virtual void CryptoPwhashStrVerifyAsync::HandleErrorCallback()’:
../src/crypto_pwhash_str_verify_async.cc:42:29: error: ‘async_resource’ was not declared in this scope
     callback->Call(2, argv, async_resource);
                             ^~~~~~~~~~~~~~

```

**Comment by mafintosh** [comment]
Can you try clearing your .node-gyp folder in home and try again?

**Comment by vladimiry** [comment]
Compiling for node 11 on linux x64 works for me, just checked. Also compiles well for electron@v5 (didn't try to run though).

**Comment by emilbayes** [comment]
@trxcllnt Did you try removing the .node-gyp cache Mathias mentioned?

**Comment by trxcllnt** [comment]
sorry, haven't had time to switch back to the arm machine since then, but I'll give it a shot in an hour or so.

**Comment by trxcllnt** [comment]
Just tried again after clearing out `~/.node-gyp` and I'm still getting it. I also tried with node 10 and 8 LTS just to make sure and I'm seeing the errors in those versions too. This might have something to do with the version of gcc I have on arm64. I'm running this on nvidia's [jetson xavier](https://www.developer.nvidia.com/embedded/buy/jetson-agx-xavier-devkit), and the tegra cuda has some special requirements WRT to gcc/g++.

**Comment by emilbayes** [comment]
I don't think there's much we can do here. Judging from the errors it looks like it is using the wrong header files. If you find a way to resolve this I'm happy to take a PR :)

**Comment by christianbundy** [comment]
For future bug-hunters: I was able to reproduce this with sodium-native@2.2.2 but the issue seems to be resolved with sodium-native@2.3.0. Unfortunately `npm update --depth 999 sodium-native` wasn't working so I had to `rm -rf node_modules package-lock.json && npm install`. Godspeed.
